File: frontend/src/api/rosetta.ts
import { api, apiJson } from './client'
import type { RosettaEvalJob, RosettaEvalJobListOut } from './types'

export type RosettaEvalCreateBody = {
  name?: string | null
  wt_fold_job_id?: string | null
  mutant_fold_job_ids?: string[]
  nstruct?: number
  n_jobs?: number
  antibody_chains?: string
  antigen_chains?: string
}

export async function fetchRosettaEvalJobs(limit = 50) {
  return apiJson<RosettaEvalJobListOut>(`/api/rosetta-eval-jobs?limit=${limit}`)
}

export async function fetchRosettaEvalJob(id: string) {
  return apiJson<RosettaEvalJob>(`/api/rosetta-eval-jobs/${id}`)
}

export async function createRosettaEvalJob(body: RosettaEvalCreateBody) {
  return apiJson<RosettaEvalJob>('/api/rosetta-eval-jobs', { method: 'POST', data: body })
}

export async function uploadRosettaEvalJob(
  wt: File,
  mutants: File[],
  body: RosettaEvalCreateBody,
) {
  const fd = new FormData()
  fd.append('wt', wt)
  mutants.forEach((file) => fd.append('mutants', file))
  if (body.name) fd.append('name', body.name)
  fd.append('nstruct', String(body.nstruct ?? 3))
  fd.append('n_jobs', String(body.n_jobs ?? 16))
  fd.append('antibody_chains', body.antibody_chains ?? '')
  fd.append('antigen_chains', body.antigen_chains ?? '')
  return apiJson<RosettaEvalJob>('/api/rosetta-eval-jobs/upload', { method: 'POST', data: fd })
}

export async function deleteRosettaEvalJob(id: string) {
  await apiJson(`/api/rosetta-eval-jobs/${id}`, { method: 'DELETE' })
}

export async function downloadRosettaEvalFile(id: string, filename: string) {
  const response = await api.get(`/api/rosetta-eval-jobs/${id}/files/${encodeURIComponent(filename)}`, {
    responseType: 'blob',
  })
  const url = URL.createObjectURL(response.data)
  const anchor = document.createElement('a')
  anchor.href = url
  anchor.download = filename
  anchor.style.display = 'none'
  document.body.appendChild(anchor)
  anchor.click()
  window.setTimeout(() => {
    URL.revokeObjectURL(url)
    anchor.remove()
  }, 1000)
}

File: app/schemas/rosetta.py
"""Rosetta evaluation schemas."""

from pydantic import BaseModel, Field

from .common.jobs import JobOut


class RosettaEvalJobCreate(BaseModel):
    name: str | None = Field(default=None, max_length=128)
    wt_fold_job_id: str | None = None
    mutant_fold_job_ids: list[str] = Field(default_factory=list)
    nstruct: int = Field(default=3, ge=1, le=10)
    n_jobs: int = Field(default=16, ge=1, le=64, description="并行 CPU 进程数")
    antibody_chains: str = Field(default="", max_length=32, description="空格分隔，空则自动识别")
    antigen_chains: str = Field(default="", max_length=32)


class RosettaEvalJobOut(JobOut):
    pass


class RosettaEvalJobListOut(BaseModel):
    items: list[RosettaEvalJobOut]
    total: int
